File: leetcode/lc-us/cpp/2467.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int mostProfitablePath(vector<vector<int>> &edges, int bob, vector<int> &amount) {
        int n = (int) amount.size();
        vector<vector<int>> g(n);
        for (auto &e: edges) {
            int x = e[0], y = e[1];
            g[x].push_back(y);
            g[y].push_back(x);
        }

        int p[n];
        function<void(int, int)> dfs1 = [&](int u, int parent) {
            p[u] = parent;
            for (int x: g[u]) {
                if (x == parent) continue;
                dfs1(x, u);
            }
        };
        dfs1(0, -1);

        int f1[n];
        memset(f1, 0, sizeof f1);
        for (int step = 1; bob != -1; step++) {
            f1[bob] = step;
            bob = p[bob];
        }

        bool st[n];
        memset(st, false, sizeof st);
        int res = INT32_MIN;
        function<void(int, int, int)> dfs2 = [&](int u, int step, int w) {
            st[u] = true;
            int v = amount[u];
            if (step == f1[u]) v /= 2;
            else if (f1[u] && f1[u] < step) v = 0;
            bool flag = true;
            for (int x: g[u]) {
                if (st[x]) continue;
                dfs2(x, step + 1, w + v);
                flag = false;
            }
            if (flag) res = max(res, w + v);
        };
        dfs2(0, 1, 0);
        return res;
    }
};

File: leetcode/lc-us/py/2467.py
from collections import defaultdict

from typing import List


class Solution:
    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        g = defaultdict(list)
        for e in edges:
            x, y = e
            g[x].append(y)
            g[y].append(x)

        parent = defaultdict(int)

        def dfs1(u: int, p: int) -> None:
            if p != -1:
                parent[u] = p
            for v in g[u]:
                if v != p:
                    dfs1(v, u)

        dfs1(0, -1)

        nodes_on_path = defaultdict(int)

        step = 0
        while bob:
            nodes_on_path[bob] = step
            step += 1
            bob = parent[bob]
        nodes_on_path[0] = step

        res = -10 ** 10

        def dfs2(u: int, p: int, w: int) -> None:
            t = 0
            if u in nodes_on_path:
                if len(nodes_on_path) & 1 and nodes_on_path[u] == nodes_on_path[0] // 2:
                    t = amount[u] // 2
                elif nodes_on_path[u] > nodes_on_path[0] // 2:
                    t = amount[u]
            else:
                t = amount[u]
            nonlocal res
            if len(g[u]) == 1 and u:
                res = max(res, w + t)
            for v in g[u]:
                if v != p:
                    dfs2(v, u, w + t)

        dfs2(0, -1, 0)

        return res
